### DISCUSSION TITLE
Multiple dispatching rules

### DISCUSSION THREAD

**Discussion by juaninicolai** [microcks/microcks#1258] [Q&A] [answered]
Hello, does someone know if it's possible to configure multiple dispatching rules for one mocked endpoint?

I am using JSON_BODY rule to infer which response (out of 2 options) to return.
After reading this proxy [article](https://microcks.io/blog/new-proxy-features-1.9.1/) I thought of implementing this for the same endpoint.

The implementation would be as follows (this is working fine):
```
{
  "exp": "/destinationLocation/rkstCode",
  "operator": "equals",
  "cases": {
    "DKAAR": "dkaar",
    "NLROT": "nlrot",
    "default": "Destination not found"
  }
}
```

But when trying to add the PROXY_FALLBACK configuration via UI and saving it, in the details page, the JSON_BODY rule is gone and replaced by the latter.

The configured proxy rules is this one (not working):
```
{
  "dispatcher": "JSON_BODY",
  "dispatcherRules": {
    "exp": "/destinationLocation/rkstCode",
    "operator": "equals",
    "cases": {
      "DKAAR": "dkaar",
      "NLROT": "nlrot",
      "default": "Destination not found"
    }
  },
  "proxyUrl": URL_HERE
}
```

And when hitting the endpoint I get this error from the server: `Given JSON string cannot be interpreted as valid ProxyFallbackSpecification` which makes me wonder if the issue is related to multiple rules or just JSON_BODY not being allowed to be wrapped by PROXY_FALLBACK.

**Comment by lbroudoux** [comment]
Have you tried converting the `dispatcherRules` as a String? (putting the content within "" and escaping stuffs?
I admit I never tried this but `dispatcherRules` being defined as String, it should work...
It may be more convenient in the future to extend the type to `String | Node`.
Let us know.

**Reply by juaninicolai** [reply]
Hi Laurent,

Yes, I tried passing the dispatcherRule as a String and got this error from the server: io.github.microcks.util.dispatcher.JsonMappingException: Given JSON string cannot be interpreted as valid ProxyFallbackSpecificatio at io.github.microcks.util.dispatcher.ProxyFallbackSpecification.buildFromJsonString(ProxyFallbackSpecification.java:68) so it's not working at the moment.

I will continue trying other approaches and let you know.

**Reply by juaninicolai** [reply, marked as answer]
Hey @lbroudoux, this version worked half-way through:

```
{
  "dispatcher": "JSON_BODY",
  "dispatcherRules": "{\n    \"exp\": \"/destinationLocation/rkstCode\",\n    \"operator\": \"equals\",\n    \"cases\": {\n      \"DKAAR\": \"dkaar\",\n      \"NLROT\": \"nlrot\",\n      \"default\": \"Destination not found\"\n    }\n  }",
  "proxyUrl": "URL_HERE"
}
```

I am gonna keep working on this and keep you posted :)
Thanks

**Reply by lbroudoux** [reply]
Are you still getting an exception here? Or does it fail somewhere else?

**Reply by juaninicolai** [reply]
Hey, no exception at the moment but I am getting a BAD_REQUEST from the mocked API, which is unexpected. 
I sent a request with a non-handled value in the JSON_BODY rule, which I thought would trigger the 'default' response OR the proxy to redirect to the production API but that did not happen.

I am still working on this, will update this thread when this is working or new questions arise :)

**Reply by juaninicolai** [reply]
Now it works. It seems the 'default' rule of JSON_BODY gets disabled or does not apply when wrapped via proxy.
There was nothing wrong with the proxy dispatcher, just a wrong implementation from my side (but I learned a lot). 
The above JSON rule (with strings and escaped fields) works OK.
